### PR TITLE
ARM/AArch64 CPU host detection

### PR DESCRIPTION
**Pull request by LemonBoy**
Please give this a spin on your Linux ARM boxes and post the output of `zig build-obj --show-builtin` (together with your `/proc/cpuinfo` if the detection fails)

cc @joachimschmidt557 
cc @mikdusan I think we're missing some entries for some of the CI boxes, do you still have their cpuinfo ?

### CODE REVIEW

**Comment by mikdusan** [comment]
You're looking for both `lscpu` and `/proc/cpuinfo` right? I only captured the `lscpu` output from cavium and falkor. Here's a grab from an arm server that is not in drone ci's pipeline pool:

```
[mike@arm64-test:~]$ lscpu
Architecture:        aarch64
Byte Order:          Little Endian
CPU(s):              32
On-line CPU(s) list: 0-31
Thread(s) per core:  1
Core(s) per socket:  32
Socket(s):           1
NUMA node(s):        1
Vendor ID:           APM
Model:               2
Model name:          X-Gene
Stepping:            0x3
CPU max MHz:         3300.0000
CPU min MHz:         363.9700
BogoMIPS:            90.00
L1d cache:           32K
L1i cache:           32K
L2 cache:            256K
NUMA node0 CPU(s):   0-31
Flags:               fp asimd evtstrm aes pmull sha1 sha2 crc32 cpuid

[mike@arm64-test:~]$ head -20 /proc/cpuinfo
processor	: 0
BogoMIPS	: 90.00
Features	: fp asimd evtstrm aes pmull sha1 sha2 crc32 cpuid
CPU implementer	: 0x50
CPU architecture: 8
CPU variant	: 0x3
CPU part	: 0x000
CPU revision	: 2
```

**Comment by joachimschmidt557** [comment]
Here is the output of `zig build-obj --show-builtin` (using LLVM to detect the CPU) and `cat /proc/cpuinfo` from Cavium ThunderX 88XX:

```zig
usingnamespace @import("std").builtin;
/// Deprecated
pub const arch = Target.current.cpu.arch;
/// Deprecated
pub const endian = Target.current.cpu.arch.endian();

/// Zig version. When writing code that supports multiple versions of Zig, prefer
/// feature detection (i.e. with `@hasDecl` or `@hasField`) over version checks.
pub const zig_version = try @import("std").SemanticVersion.parse("0.8.0-dev.2082+e9e91b4ed");
pub const zig_is_stage2 = true;

pub const output_mode = OutputMode.Obj;
pub const link_mode = LinkMode.Static;
pub const is_test = false;
pub const single_threaded = false;
pub const abi = Abi.gnu;
pub const cpu: Cpu = Cpu{
    .arch = .aarch64,
    .model = &Target.aarch64.cpu.thunderxt88,
    .features = Target.aarch64.featureSet(&[_]Target.aarch64.Feature{
        .aes,
        .crc,
        .crypto,
        .fp_armv8,
        .neon,
        .perfmon,
        .predictable_select_expensive,
        .sha2,
        .use_postra_scheduler,
        .v8a,
    }),
};
pub const os = Os{
    .tag = .linux,
    .version_range = .{ .linux = .{
        .range = .{
            .min = .{
                .major = 4,
                .minor = 19,
                .patch = 0,
            },
            .max = .{
                .major = 4,
                .minor = 19,
                .patch = 0,
            },
        },
        .glibc = .{
            .major = 2,
            .minor = 17,
            .patch = 0,
        },
    }},
};
pub const object_format = ObjectFormat.elf;
pub const mode = Mode.Debug;
pub const link_libc = false;
pub const link_libcpp = false;
pub const have_error_return_tracing = true;
pub const valgrind_support = false;
pub const position_independent_code = false;
pub const position_independent_executable = false;
pub const strip_debug_info = false;
pub const code_model = CodeModel.default;
```


```
processor	: 0
BogoMIPS	: 200.00
Features	: fp asimd evtstrm aes pmull sha1 sha2 crc32 cpuid
CPU implementer	: 0x43
CPU architecture: 8
CPU variant	: 0x1
CPU part	: 0x0a1
CPU revision	: 1
```

File: lib/std/Thread.zig [hunk 1]
```diff
@@ -75,7 +75,7 @@ pub fn spinLoopHint() callconv(.Inline) void {
         },
         .arm, .armeb, .thumb, .thumbeb => {
             // `yield` was introduced in v6k but are also available on v6m.
-            const can_yield = comptime std.Target.arm.featureSetHas(std.Target.current.cpu.features, .has_v6m);
+            const can_yield = comptime std.Target.arm.featureSetHasAny(std.Target.current.cpu.features, .{ .has_v6k, .has_v6m });
             if (can_yield) asm volatile ("yield" ::: "memory");
```
Comment (xackus) [inline lib/std/Thread.zig:79]: Shouldn't this fallback to `asm volatile ("" ::: "memory")`?

**Comment by jedisct1** [comment]
Detection seem to be fine here (Cortex-A72)

```zig
usingnamespace @import("std").builtin;
/// Deprecated
pub const arch = Target.current.cpu.arch;
/// Deprecated
pub const endian = Target.current.cpu.arch.endian();

/// Zig version. When writing code that supports multiple versions of Zig, prefer
/// feature detection (i.e. with `@hasDecl` or `@hasField`) over version checks.
pub const zig_version = try @import("std").SemanticVersion.parse("0.8.0-dev.2092+4f9c928e5");
pub const zig_is_stage2 = true;

pub const output_mode = OutputMode.Obj;
pub const link_mode = LinkMode.Static;
pub const is_test = false;
pub const single_threaded = false;
pub const abi = Abi.gnu;
pub const cpu: Cpu = Cpu{
    .arch = .aarch64,
    .model = &Target.aarch64.cpu.cortex_a72,
    .features = Target.aarch64.featureSet(&[_]Target.aarch64.Feature{
        .aes,
        .crc,
        .crypto,
        .fp_armv8,
        .fuse_aes,
        .fuse_literals,
        .neon,
        .perfmon,
        .sha2,
        .v8a,
    }),
};
pub const os = Os{
    .tag = .linux,
    .version_range = .{ .linux = .{
        .range = .{
            .min = .{
                .major = 5,
                .minor = 11,
                .patch = 0,
            },
            .max = .{
                .major = 5,
                .minor = 11,
                .patch = 0,
            },
        },
        .glibc = .{
            .major = 2,
            .minor = 33,
            .patch = 0,
        },
    }},
};
pub const object_format = ObjectFormat.elf;
pub const mode = Mode.Debug;
pub const link_libc = false;
pub const link_libcpp = false;
pub const have_error_return_tracing = true;
pub const valgrind_support = false;
pub const position_independent_code = false;
pub const position_independent_executable = false;
pub const strip_debug_info = false;
pub const code_model = CodeModel.default;
```

File: lib/std/zig/system/linux.zig [hunk 2]
```diff
@@ -131,23 +131,232 @@ const PowerpcCpuinfoImpl = struct {
...
+            self.have_fields += 1;
+        } else if (mem.eql(u8, key, "CPU variant")) {
+            info.variant = try fmt.parseInt(u8, value, 0);
+            self.have_fields += 1;
+        } else if (mem.eql(u8, key, "CPU part")) {
+            info.part = try fmt.parseInt(u16, value, 0);
+            self.have_fields += 1;
+        } else if (mem.eql(u8, key, "model name")) {
+            // ARMv6 cores report "CPU architecture" equal to 7.
+            if (mem.indexOf(u8, value, "(v6l)")) |_| {
```
Comment (Mouvedia) [inline lib/std/zig/system/linux.zig:269]: ref: https://issuetracker.google.com/issues/36920862

**Comment by Mouvedia** [comment]
@LemonBoy check https://android.googlesource.com/platform/external/cpu_features/+/15181ac63e8f985f645d2445fff64948521d5496/src/cpuinfo_arm.c#121

**Comment by daurnimator** [comment]
> Please give this a spin on your Linux ARM boxes and post the output of `zig build-obj --show-builtin` (together with your `/proc/cpuinfo` if the detection fails)

via Works-on-arm we currently have access to these CPUs:
  - 2x Cavium ThunderX CN8890 @2GHz
  - 2x 32-core HiSilicon Hi1616 @ 2.4Ghz
  - 1x Ampere eMAG 8180 32-core @ 3.0Ghz

Would any of these be helpful?

File: lib/std/Thread.zig (hunk 1 above)
Comment (LemonBoy) [reply lib/std/Thread.zig]: Good catch (even though I really doubt anyone is going to use that code path, heh).

**Comment by LemonBoy** [comment]
I've added a few more models from several vendors so that we should be able to cover many more systems.

Some models (eg. Ampere eMAG or APM X-Gene) are not recognized by LLVM. We can solve this problem by adding a model [here](https://github.com/ziglang/zig/blob/master/tools/update_cpu_features.zig) and updating the detection logic.

Exercise for the reader:
* Add both the aforementioned CPU models to `update_cpu_features.zig` (Use [this](https://github.com/gcc-mirror/gcc/blob/master/gcc/config/aarch64/aarch64-cores.def#L74) as reference to fill in the feature flags)
* Update the feature detection code to recognize vendor 0x50 (APM)
* Add the two missing CPUs (note that they only differ by the variant number)

File: lib/std/zig/system/linux.zig [hunk 3]
```diff
@@ -163,61 +163,113 @@ const ArmCpuinfoImpl = struct {
...
+        const HiSilicon = [_]E{
+            E{ .part = 0xd01, .m64 = &A64.tsv110 },
+        };
+        // implementer = 0x4e
+        const Nvidia = [_]E{
+            E{ .part = 0x004, .m64 = &A64.carmel },
+        };
+        // implementer = 0x51
+        const Qualcomm = [_]E{
+            E{ .part = 0x06f, .m32 = &A32.krait },
```
Comment (Mouvedia) [inline lib/std/zig/system/linux.zig:244]: Don't you have to add idiva and idivt to its features?
cf https://android.googlesource.com/platform/external/cpu_features/+/15181ac63e8f985f645d2445fff64948521d5496/src/cpuinfo_arm.c#139

Comment (LemonBoy) [reply lib/std/zig/system/linux.zig:244]: That has nothing to do with what we're doing here, the feature set in `lib/std/target/arm.zig` is correct.

**Review by andrewrk — APPROVED** [review]
looks great, are we waiting on anything?

**Comment by LemonBoy** [comment]
It's good to go, we can probably unhook the LLVM feature detection and rely on the native Zig one!

This PR closes #5110 :partying_face:

**Pull request merged**